### PR TITLE
Replace all references to ReleaseType with Release

### LINKED ISSUE #5: Releases endpoint (https://pdc.fedoraproject.org/rest_api/v1/releases/) - Database Model
Release specification

Attribute name | Type | Example
-- | -- | --
release_id | string | ‘fedora-28’
short | string | ‘fedora’
version | integer | 28
name | string | ‘Fedora’
release_date | string | “2018-09-03”
eol_date | string | “2018-09-04”
release_type | ReleaseType.short | "development" when today () < release_date"eol" when today() >= eol_date"ga" when  today() > release_date and today() < eol_date
active | boolean | true when release_type is ga
sigkey | string | ‘81b46521’

ReleaseType  specification

Attribute name | Type | Example
-- | -- | --
short | string | ‘ga’
name | string | ‘Release’
suffix | string | ‘-updates-testing’

Condition of Statisfaction:
- [ ] New DRF application created `releases`
- [ ] Release and ReleaseType model done.

### PR DESCRIPTION
**Pull request by subpop**
The `Release` model type has two annotated properties: `release_type` and `active`. They are annotated onto a custom `ModelManager` subclass so they can be used in database queries (i.e. `filter()`).

## Note ##

This PR also replaces the original `0001_create_release_type.py` migration with a new `0001_initial.py`. Rather than adding two new migrations (one to create the `Release` type and a second one to remove `ReleaseType`).

fixes #5, #10

### CODE REVIEW

**Review by cverna — COMMENTED** [review]
Sorry to be a pain, but the more I think about it, I think we should have the current `release_type` just called `type` and be simple text field where we can use values like `update` or `ga`.

Then the active boolean should still be computed but based on the dates instead of based on current `release_type`. ie active False when date < release_date or date > eol_date else True.

The the API consumer can just use the active=True or active=False to query the releases.

What do you think ? 

Also I think the `type` and `active` field should be displayed in the GET request result.

**Comment by cverna** [comment]
CC'ing @abompard since he looked at this in pdc.

**Comment by subpop** [comment]
> Sorry to be a pain, but the more I think about it, I think we should have the current `release_type` just called `type` and be simple text field where we can use values like `update` or `ga`.

That seems reasonable. Are you thinking it should be a stored property as a column in the database then? I would reconsider the name though. "type" is often a protected keyword in many programming languages. In Python in particular, `type()` is a global function to return the dynamic type of an object at runtime. As a property on an object, it's *probably* safe; but as a rule, I try to avoid using protected keywords like that.

> Then the active boolean should still be computed but based on the dates instead of based on current `release_type`. ie active False when date < release_date or date > eol_date else True.
>
> The the API consumer can just use the active=True or active=False to query the releases.
> 
> What do you think ?

These calculated properties are quite unwieldy. They require a customized model `Manager` subclass. I suspect they will also require overriding some methods on the custom `ModelSerializer` subclass. Another approach we could take is this. Add them as regular stored fields on the `Release` class. Then for the properties that we want to be "computed", we mark them as read-only, so no write methods will attempt to change their value. Whenever the value of a property the field derives from is changed, recompute the derived value and store it. I know how to do this in languages like Objective-C or Java with their custom setters and getters, but Django models are a whole different beast.

*Another* option is to support "active" as a view-level query parameter. Then the logic to find what records are considered "active" can exist inside the view handler.

> Also I think the `type` and `active` field should be displayed in the GET request result.

This is probably an oversight on my part. They need to be added as fields into the `ReleaseSerializer`. I'll figure out how to add them.

**Comment by cverna** [comment]
> > Sorry to be a pain, but the more I think about it, I think we should have the current `release_type` just called `type` and be simple text field where we can use values like `update` or `ga`.
> 
> That seems reasonable. Are you thinking it should be a stored property as a column in the database then? I would reconsider the name though. "type" is often a protected keyword in many programming languages. In Python in particular, `type()` is a global function to return the dynamic type of an object at runtime. As a property on an object, it's _probably_ safe; but as a rule, I try to avoid using protected keywords like that.
> 

Yes I think a stored property will do, maybe we could use `status` instead of `type`. I am trying to think about the api user and I am usually annoyed when I have overly long name like `release_type` 

> > Then the active boolean should still be computed but based on the dates instead of based on current `release_type`. ie active False when date < release_date or date > eol_date else True.
> > The the API consumer can just use the active=True or active=False to query the releases.
> > What do you think ?
> 
> These calculated properties are quite unwieldy. They require a customized model `Manager` subclass. I suspect they will also require overriding some methods on the custom `ModelSerializer` subclass. Another approach we could take is this. Add them as regular stored fields on the `Release` class. Then for the properties that we want to be "computed", we mark them as read-only, so no write methods will attempt to change their value. Whenever the value of a property the field derives from is changed, recompute the derived value and store it. I know how to do this in languages like Objective-C or Java with their custom setters and getters, but Django models are a whole different beast.

Yes I need to read more about the `Manager` and `ModelSerializer`.
> 
> _Another_ option is to support "active" as a view-level query parameter. Then the logic to find what records are considered "active" can exist inside the view handler.

I think as the user of the API I would have 2 use cases, one is to query all the active releases something like `/api/v1/releases?active=true` which the view-level query parameter would provide but I can imagine another use case which would be to query a specific version for example f29 `/api/v1/releases/?version=29` and check if it is active.

> 
> > Also I think the `type` and `active` field should be displayed in the GET request result.
> 
> This is probably an oversight on my part. They need to be added as fields into the `ReleaseSerializer`. I'll figure out how to add them.
Cool 

Thanks for working on this, your PR looks really good it is just that I am trying to sort out the specification :monkey_face:

**Comment by subpop** [comment]
> Yes I think a stored property will do, maybe we could use `status` instead of `type`. I am trying to think about the api user and I am usually annoyed when I have overly long name like `release_type`

Me too. I like `status`, or `state` or `stage`?

> I think as the user of the API I would have 2 use cases, one is to query all the active releases something like `/api/v1/releases?active=true` which the view-level query parameter would provide but I can imagine another use case which would be to query a specific version for example f29 `/api/v1/releases/?version=29` and check if it is active.

Oh yea good point. We'd definitely want that in the data returned by a query like that.

**Comment by subpop** [comment]
Actually, storing `release_type` and updating it only when `eol_date` and/or `release_date` change won't work. Because `datetime.date.today()` changes every day. ^_^ It has to be calculated at the time of retrieval. But it looks like adding calculated fields to a serializer is trivial.

**Comment by subpop** [comment]
I think this current implementation is a misuse of `annotate()`. While it *does* work, it does not seem to be the idiomatic way of creating calculated fields on a `Model`. It makes the calculated field available through `QuerySet` filters, but it does not add attributes on individual model instances, therefore it does not include the attribute values during serialization. When I suggested using a calculated field, I had no idea it would be this complex; require this many touch points inside the various layers (models, serializers, views). I'll have to think more about this.

**Comment by cverna** [comment]
> I think this current implementation is a misuse of `annotate()`. While it _does_ work, it does not seem to be the idiomatic way of creating calculated fields on a `Model`. It makes the calculated field available through `QuerySet` filters, but it does not add attributes on individual model instances, therefore it does not include the attribute values during serialization. When I suggested using a calculated field, I had no idea it would be this complex; require this many touch points inside the various layers (models, serializers, views). I'll have to think more about this.

Maybe we can just create the base fields in this PR. (excluding `status` and `active`) and deal with these 2 in another PR. Up to you :small_airplane:

**Comment by subpop** [comment]
I found a package, [django-computed-property](https://github.com/brechin/django-computed-property) that solves this exact problem. Implementing these two fields as `ComputedCharField` and `ComputedBooleanField` implement the exact behavior I was hoping for.

File: fpdc/releases/migrations/0001_initial.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+                ),
+                ("release_id", models.CharField(max_length=255, unique=True)),
+                ("short", models.CharField(max_length=255)),
+                ("version", models.IntegerField()),
+                ("name", models.CharField(max_length=255)),
+                ("release_date", models.DateField()),
+                ("eol_date", models.DateField()),
+                ("sigkey", models.CharField(max_length=255)),
+                (
+                    "release_type",
```
Comment (cverna) [inline fpdc/releases/migrations/0001_initial.py]: So do we want to change that to `status` or `state` ? I let you choose :+1:

File: fpdc/releases/models.py [hunk 2]
```diff
@@ -1,8 +1,30 @@
...
+    name = models.CharField(max_length=255, blank=False, unique=False)
+    release_date = models.DateField()
+    eol_date = models.DateField()
+    sigkey = models.CharField(max_length=255, blank=False)
+    release_type = computed_property.ComputedCharField(
+        compute_from="_release_type", max_length=255, null=True
+    )
+    active = computed_property.ComputedBooleanField(compute_from="_active", default=True)
+
+    @property
```
Comment (cverna) [inline fpdc/releases/models.py]: nice :+1:

File: fpdc/releases/tests/test_models.py [hunk 3]
```diff
@@ -1,19 +1,28 @@
...
-        assert obj.name == "Release", "Should have a name field"
-        assert obj.suffix == "-updates-testing", "Should have a suffix field"
+    def test_query(self):
+        """Ensure we can query for instances."""
+        mixer.blend(Release, short="fedora")
+        r = Release.objects.get(short__exact="fedora")
+        assert r.short == "fedora"
+
+    def test_calculated_fields(self):
+        """Ensure we can query calculated fields."""
```
Comment (cverna) [inline fpdc/releases/tests/test_models.py:43]: Do we also want to test the case of a `ga` and `development` release here ?

**Review by cverna — CHANGES_REQUESTED** [review]
Looks good, just a couple of comments before merging it :fireworks:

**Comment by cverna** [comment]
I have also opened a [PR](https://github.com/brechin/django-computed-property/pull/4) to django-computed-property  to add their license to the setup.py that way our license check test will be happy.

**Review by cverna — APPROVED** [review]
LGTM :+1:

**Pull request merged**